scrapy_azure_exporter/azure_exporter.py

```diff
@@ -1,45 +1,113 @@
-from azure.storage.blob import BlobServiceClient
-from azure.core.exceptions import ResourceNotFoundError
+import logging
+import os
 import typing
+from urllib.parse import urlparse
+
+from azure.core.exceptions import AzureError
+from azure.storage.blob import BlobServiceClient
+from azure.storage.blob._models import BlobType
 from scrapy.exceptions import NotConfigured
 from scrapy.extensions.feedexport import BlockingFeedStorage
-import logging
 
-"""The pipeline for adding a file to Azure Blob"""
 
+class AzureFeedStorage(BlockingFeedStorage):
+    """This class will be used for exporting files to Azure Blob Storage"""
 
-class AzureFeedExporter(BlockingFeedStorage):
+    def __init__(
+        self,
+        uri,
+        connection_string,
+        account_url_with_sas_token,
+        account_url,
+        account_key,
+        *,
+        feed_options=None,
+    ):
+        if feed_options is None:
+            feed_options = {}
+        self.connection_string = connection_string
+        self.account_url_with_sas_token = account_url_with_sas_token
+        self.account_url = account_url
+        self.account_key = account_key
+        self.overwrite = feed_options.get("overwrite", False)
+        self.blob_type = feed_options.get("blob_type", "BlockBlob")
```

Comment:
Given PageBlob does not seem to make sense for this plugin (the plugin cannot use random access), why don’t we remove the `blob_type` feed option altogether, and use `BlockBlob` and `AppendBlob` based on the value of `overwrite`?

Also, I see you used contradicting defaults here. We are free to choose whichever default value we want for `overwrite`, so if `BlockBlob` should be the default (and [it probably should](https://docs.microsoft.com/en-us/python/api/azure-storage-blob/azure.storage.blob.blobclient?view=azure-python#azure-storage-blob-blobclient-upload-blob)), `overwrite` should be `True` by default. We just need to mention that in the README.

Reply:
Thanks @Gallaecio , I agree with your suggestions.

Actually, I had a scenario in my mind that is why I chose to set those default values.
* Suppose If a person doesn't want to overwrite files by accident,
	* When using BlockBlob overwrite=False as default seemed to be a safe option.
	* When using AppendBlob, it's obvious that AppendBlob is used to append data to the existing file, so overwrite=False seemed to be obvious as well. if they really want to overwrite, they can set it overwrite=True

Let me know if that scenario is not a priority, then we can shape it according to your suggestion.

Reply:
I am fine with making `overwrite` `False` by default. It is actually consistent with Scrapy, where it is only `True` by default in scenarios where appending is not an option.

Reply:
OK, I think I did not get it before but hopefully I get it now.

So I see 3 potentially-desirable behaviors:
- write if the file does not exist, do nothing if the file exists (`x` mode of [Python’s open()](https://docs.python.org/3/library/functions.html#open))
- overwrite, even if the file exists (`w`)
- append (or write if the file does not exist yet) (`a`)

I think `x` may be a valid use case, and I am wondering if we should deprecate `overwrite` in Scrapy itself in favor of an option that allows for these 3 export modes. And it seems like it would be the safest value as well.

You seem to want to have it as a default value here. I am not a fan of users having to have 2 variables here, 1 of them being Azure-specific, when I hope we can eventually have a single variable instead. But until we have something like that available on Scrapy itself, I think it is OK to implement it here the way you propose.

I would only ask that you make it clear in the documentation what the default _mode_ is, and how to achieve any of the other 2 modes. That is, first describe the default value of `overwrite` (no need to provide more details than that, I think, as it is documented upstream and quite self-explanatory), the describe the plugin-specific `blob_type` option and its possible values (I would either omit the Page one or explain that it works the same as Blob for feed exports), and then explain the main 3 combinations of those 2 parameters (overwrite=False + Blob, overwrite=False + Append, overwrite=True).

Reply:
Thanks @Gallaecio,
I like the way you have summarized this and have linked this with those potentially-desirable behaviors. I have added the description explaining those combinations.